### ISSUE TITLE
PathHierarchy vs path_hierarchy

### ISSUE THREAD

**Issue by kstaken** [elastic/elasticsearch#15756]
In versions of Elasticsearch prior to 2.0 the documentation for the path_hierarchy analyzer says the 'type' field has these requirements. 

```
Required. Must be set to PathHierarchy (case-sensitive).
```

https://www.elastic.co/guide/en/elasticsearch/reference/1.7/analysis-pathhierarchy-tokenizer.html

In 2.0+ that is now incorrect and if you have an index that uses type: 'PathHierarchy' the upgrade process from 1.7.* to 2.* fails on each shard of the index. It requires it to be type: 'path_hierarchy'. 

We have a user that has an index with many billions of records in it that uses this analyzer as specified by the 1.7 documentation. This index is so large it is infeasible to reindex to simply change this part of the mapping. Is this an intentional change in 2.*? Is there any way to patch the mapping attached to the shards so that an upgrade to 2.* is possible?

Note: this issue is not flagged by the migration plugin. We were testing the upgrade on a test cluster and only discovered the issue when the shards attempted to upgrade. It just logs the following stack repeatedly for each shard and effectively corrupts the index since other indices had upgraded successfully while this type fails. In that scenario the cluster can't be reverted without data loss.

```
[2015-12-14 18:42:56,825][WARN ][indices.cluster          ] [node] [[logs-2015.12.11][3]] marking and sending shard failed due to [failed to create index]
[logs-2015.12.11] IndexCreationException[failed to create index]; nested: IllegalArgumentException[Unknown Tokenizer type [PathHierarchy] for [field_tokens]];
	at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:362)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.applyNewIndices(IndicesClusterStateService.java:307)
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.clusterChanged(IndicesClusterStateService.java:176)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:494)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:231)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:194)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: Unknown Tokenizer type [PathHierarchy] for [field_tokens]
	at org.elasticsearch.index.analysis.AnalysisModule.configure(AnalysisModule.java:267)
	at org.elasticsearch.common.inject.AbstractModule.configure(AbstractModule.java:61)
	at org.elasticsearch.common.inject.spi.Elements$RecordingBinder.install(Elements.java:233)
	at org.elasticsearch.common.inject.spi.Elements.getElements(Elements.java:105)
	at org.elasticsearch.common.inject.InjectorShell$Builder.build(InjectorShell.java:143)
	at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:99)
	at org.elasticsearch.common.inject.InjectorImpl.createChildInjector(InjectorImpl.java:159)
	at org.elasticsearch.common.inject.ModulesBuilder.createChildInjector(ModulesBuilder.java:55)
	at org.elasticsearch.indices.IndicesService.createIndex(IndicesService.java:358)
	... 8 more
```

**Comment by ywelsch** [comment]
This is indeed a bug. We will provide a fix for the next release. If you can't wait for the next release, you might try the following before upgrading to 2.x (try it first on a test cluster, and make backups before proceeding): Close indices. Update settings of indices by changing PathHierarchy to path_hierarchy. Open indices. That should be it. You can then upgrade to 2.x.

Example (tested on 1.7.4):

```

# This is the example index we start with
curl -X POST http://localhost:9200/myindex/ -d'
{
  "settings": {
    "analysis": {
      "analyzer": {
        "path-analyzer": {
          "type": "custom",
          "tokenizer": "path-tokenizer"
        }
      },
      "tokenizer": {
        "path-tokenizer": {
          "type": "PathHierarchy",
          "delimiter": "."
        }
      }
    }
  },
  "mappings": {
    "test": {
      "properties": {
        "text": {
          "type": "string",
          "analyzer": "path-analyzer"
        }
      }
    }
  }
}'

# Close index
curl -X POST http://localhost:9200/myindex/_close

# Update settings by changing PathHierarchy to path_hierarchy
curl -X PUT http://localhost:9200/myindex/_settings -d'
{
  "settings": {
    "analysis": {
      "analyzer": {
        "path-analyzer": {
          "type": "custom",
          "tokenizer": "path-tokenizer"
        }
      },
      "tokenizer": {
        "path-tokenizer": {
          "type": "path_hierarchy",
          "delimiter": "."
        }
      }
    }
  }
}'

# reopen index
curl -X POST http://localhost:9200/myindex/_open
```

**Comment by ywelsch** [comment]
Fixed in #15785
